Gui refactor

Namlev1 opened the pull request:
Kompletny refactor data_loader oraz gui. Wprowadzone ulepszenia tu i tam.

Nowe moduły:
- `gui_utils` - funkcje typu "utwórz przycisk", "wyświetl okienko modalne" itd.
- `processing_pipeline` - wyeksportowałem tam całą funkcję `process_bands_and_calculate_indices` żeby nie siedziało w module GUI
- `data_types.h` - definicja podstawowych struktur (~~cholerne~~ kochane C nie ma klas)

`resample` nie eksportuje już 3 różnych funkcji, tylko jedną która konwertuje wszystkie pasma do danej rozdzielczości.
`index_calculator` ma liczenie czasu (spójne z pozostałymi modułami).

Nie zmieniłem nazw ani parametrów funkcji, które były deklarowane jako API w plikach `.h` (np. `LoadBandData` w `data_loader`, które jest używane w innych modułach). Tam gdzie się da dałem wypisywanie do konsoli, później można to dodać do wyświetlania w GUI. 

Duże funkcje oraz funkcje API mają dokumentację Doxygen.

Namlev1 committed:
Refactor data_loader.c

Podział LoadBandData na funkcje

Namlev1 committed:
Dodanie logowania i pomiar czasu

Namlev1 committed:
Podzielenie handle_file_selection na funkcje

Namlev1 committed:
Ekstrakcja funkcji utility z gui.c do gui_utils.c

Namlev1 committed:
Eksport funkcji utility z data_loader.c do utils.c

Namlev1 committed:
Utworzenie struktury BandData

Namlev1 committed:
Ekstrakcja logiki resamplingu z gui.c do resampler.c

Namlev1 committed:
Ekstrakcja funkcji process_bands_and_calculate_indices

Docelowo ta funkcja zostanie zupełnie wyrzucona z gui.c

Namlev1 committed:
Dodanie logowania czasu trwania obliczania NDMI i NDVI, wypisywanie logów do konsoli

Namlev1 committed:
Poprawa logowania w free_index_map_data

Namlev1 committed:
Refactor wyświetlania okna mapy

Namlev1 committed:
Refaktoryzacja funkcji toggle up/downscaling oraz ndvi/ndmi na mapie

Namlev1 committed:
Poprawa logów, usunięcie zbędnych argumentów

Namlev1 committed:
Refaktoryzacja rysowania mapy

Namlev1 committed:
Posprzątanie pliku

Deklaracja funkcji, zmiana kolejności definicji

Namlev1 committed:
Wyeksportowanie show_error_dialog do gui_utils

Namlev1 committed:
Wyeksportowanie całego pipelinu pobierania danych, resamplingu, obliczania wskaźników do nowego modułu.

Namlev1 committed:
Resetowanie ustawień GUI po zamknięciu okna mapy

Namlev1 committed:
Usunięcie button z struct BandData

Namlev1 committed:
Ekstrakcja BandData i BandType do data_types.h

Namlev1 committed:
Przeniesienie funkcji z gui_utils do gui

Namlev1 committed:
Przeniesienie funkcji load_all_bands_data do data_loader, dodanie dokumentacji

Namlev1 committed:
Usunięcie band_index z struktury BandData w celu uproszczenia kodu

Namlev1 committed:
Dodanie dokumentacji

Namlev1 committed:
Dodanie dokumentacji

Haribs147 commented on processing_pipeline.c:
```diff
@@ -0,0 +1,234 @@
+    return 1;
+}
+
+static int validate_processing_result(const ProcessingResult* result)
```
W validate_processing_result w processing_pipeline.c sprawdzamy:

```c
if (!result->ndvi_data)
    {
        fprintf(stderr, "[%s] Błąd walidacji: Brak danych NDVI w wyniku.\n", get_timestamp());
        return 0;
    }

    if (!result->ndmi_data)
    {
        fprintf(stderr, "[%s] Błąd walidacji: Brak danych NDMI w wyniku.\n", get_timestamp());
        return 0;
    }
```

Później w on_rozpocznij_clicked obsługujemy to:

```c
if (!processing_result)
    {
        g_printerr("[%s] Wystąpił błąd podczas przetwarzania danych.\n", get_timestamp());
        show_error_dialog(parent_gtk_window, "Błąd podczas przetwarzania danych. Sprawdź konsolę dla szczegółów.");
        return;
    }
```

Więc zawsze aby create_map_window się otworzyło, musimy mieć ndvi data i ndmi data
Więc nie potrzebujemy tego:
```c
    if (!map_data->ndvi_data && !map_data->ndmi_data)
    {
        g_printerr("[%s] Błąd tworzenia mapy --- Brak danych NDMI / NDVI.\n", get_timestamp());
        free_index_map_data(map_data);
        return NULL;
    }
```
W create_map_window

I raczej nie potrzebujemy tego w on_draw_map_area:
```c
if (!map_data)
    {
        g_printerr("[%s] Błąd tworzenia mapy --- Brak wskaźnika do danych mapy. Rysowanie tła.\n", get_timestamp());
        cairo_set_source_rgb(cr, 0.1, 0.1, 0.1);
        cairo_paint(cr);
        return TRUE;
    }

    const float* data_to_render = strcmp(map_data->current_map_type, "NDVI") == 0
                                      ? map_data->ndvi_data
                                      : map_data->ndmi_data;
    if (!data_to_render)
    {
        g_printerr("[%s] Błąd tworzenia mapy --- Brak wskaźnika do mapy %s.\n", get_timestamp(),
                   map_data->current_map_type);
        cairo_set_source_rgb(cr, 0.2, 0.2, 0.2);
        cairo_paint(cr);
        return TRUE;
    }
```
Tu używamy co prawda map_data, ale wcześniej skopiowaliśmy po prostu s processing_result do map_data, a i tak sprawdziliśmy potem map_data czy nie jest null. Czemu nie potrzbujemy tego if'a, bo zakładamy że strcmp zawsze się powiedzie jak ma dane, a dane muszą być bo wcześniej je sprawdziliśmy. Do zastanowienia też jest czy my potrzebujemy struktury:
```c
typedef struct
{
    float* ndvi_data;
    float* ndmi_data;
    int width;
    int height;
    char current_map_type[10];
} IndexMapData;
```
Jak i tak mamy strukturę processing result, która jest praktycznie taka sama:
```c

typedef struct
{
    float* ndvi_data;
    float* ndmi_data;
    int width;
    int height;
} ProcessingResult;
```

Haribs147 commented on gui.c:
```diff
@@ -665,4 +316,291 @@ int run_gui(int argc, char *argv[]) {
+    }
+
+    // Konwersja wyniku na format GUI
+    IndexMapData* map_display_data = convert_processing_result_to_map_data(processing_result);
```
Czy my potrzebujemy tutaj struktury IndexMapData która jest tak naprawdę strukturą ProcessingResult z jednym dodatkowym polem. Większy opis w komentarzu w processing_pipeline.

Haribs147 reviewed (changes requested):
Ogólnie kodzik zajebisty, bardzo dumny jestem dobra robota. Trochę pytań zadałem w komentarzach, jak się zgadzasz z zaproponowaną zmianą to nie musisz nic pisać po prostu napraw zmianę i zrób resolve conversation, jak się nie zgadzasz to napisz i jak skończysz review napisz mi na priv. Jak są pytania w komentarzu to możesz mi odpisać albo możemy się potem zgadać na call i mi wytłumaczysz.
Ogólnie dobra robota, kodzik przepiękny <3

Namlev1 committed:
Poprawa kodu w resampler.c

Wyrzuciłem free() w replace_band_data, ponieważ ten warunek nigdy nie był spełniony. I tak zwalniamy pamięć potem w processing_pipeline

Namlev1 committed:
Usunięcie zbędnych walidacji i weryfikacji

Namlev1 committed:
Usunięcie IndexMapData

Haribs147 reviewed (approved):
Wygląda git, tą funkcje w resampler.c replace_band_data można teraz wywalić chyba bo ona tylko robi:
```c
void replace_band_data(BandData* band_data, float* new_data)
{
    *band_data->processed_data = new_data;
}
```
Ale to już w następnym commicie się zrobi
